### ISSUE TITLE
S3DownloadHandler causes 1s startup delay for people who don't use S3

### ISSUE THREAD

**Issue by kmike** [scrapy/scrapy#1054]
S3DownloadHandler tries to connect to s3 on startup, with 1s timeout. This can be seen in Scrapy logs - note 1s delay between "Enabled extensions" and "Enabled downloader middlewares" messages:

```
2015-02-27 04:49:42+0500 [scrapy] INFO: Scrapy 0.25.1 started (bot: scrapybot)
2015-02-27 04:49:42+0500 [scrapy] INFO: Optional features available: ssl, http11, boto, django
2015-02-27 04:49:42+0500 [scrapy] INFO: Overridden settings: {'DEFAULT_ITEM_CLASS': 'dirbot.items.Website', 'NEWSPIDER_MODULE': 'dirbot.spiders', 'SPIDER_MODULES': ['dirbot.spiders']}
2015-02-27 04:49:42+0500 [scrapy] INFO: Enabled extensions: LogStats, TelnetConsole, CloseSpider, CoreStats, SpiderState
2015-02-27 04:49:43+0500 [scrapy] INFO: Enabled downloader middlewares: HttpAuthMiddleware, DownloadTimeoutMiddleware, UserAgentMiddleware, RetryMiddleware, DefaultHeadersMiddleware, MetaRefreshMiddleware, HttpCompressionMiddleware, RedirectMiddleware, CookiesMiddleware, ChunkedTransferMiddleware, DownloaderStats
```

With s3 download handler commented out the delay disappears.

What do you think about making `.conn` attribute lazy? It won't be possible to raise NotConfigured exception earlier, but users will get an error later.

Another option is to disable s3 handler by default, but it is more intrusive.

**Comment by nramirezuy** [comment]
@kmike You can raise `NotConfigured` if [these](https://github.com/scrapy/scrapy/blob/master/scrapy/core/downloader/handlers/s3.py#L39) aren't present or are empty; that will fix the one second delay for most of the cases.

**Comment by kmike** [comment]
@nramirezuy a good idea

**Comment by kmike** [comment]
@nramirezuy the problem with raising NotConfigured it is that when these options are None boto tries to get them from environment variables or from config. I don't quite like an idea of replicating boto logic just to raise a NotConfigured error earlier. We can create boto.provider.Provider and use its get_credentials method to check if the credentials are set, but to me it also looks like a hack.

**Comment by nramirezuy** [comment]
@kmike Yeah I forgot about that one. The provider seems fine to me it has been there since [2012](https://github.com/boto/boto/commit/3328d99106cfc7b8773985eab086046bfdde20dd), It isn't documented but boto doesn't seems to document all the features it has, not sure whats the policy there.

**Comment by kmike** [comment]
@nramirezuy it turns out checking the credentials manually won't help because it is the credentials check which causes the delay. See https://github.com/scrapy/scrapy/pull/1055#issuecomment-81831270.

**Comment by barraponto** [comment]
I think it is ok to disable S3 by default. As long as we highlight the documentation on how to turn it on...

**Comment by nyov** [comment]
I second that, haven't used boto yet, but it's always imported at any scraper start for `optional_features` checking.

Might be an option to package all the S3-related stuff (S3FeedStorage, S3FilesStore, S3DownloadHandler) on their own?

Or how about lazy-loading `DownloadHandlers`? Would that be okay, to instance them on their first download_request for that scheme instead of always?

**Comment by kmike** [comment]
Is it fixed by https://github.com/scrapy/scrapy/pull/1421?

**Comment by nyov** [comment]
Yep. I would say so.

**Comment by gautham-nxt-gen-soft** [comment]
Hey guys I'm new to web crawling.Sorry for disturbing you guys.
I came to know that scrapy is the fastest web framework module for web crawling.
I thought that you guys would be doing that in multithreading but I heard that you guys never use any threads.How could this be possible?can anyone give a short explanation to me...Please

**Comment by stummjr** [comment]
Will the fix for this be in the next Scrapy release?

**Comment by kmike** [comment]
@stummjr yes, absolutely. Does it cause troubles for you in Scrapy 1.0?

**Comment by stummjr** [comment]
Not really. There was an user asking for help on scrapy-users because he was getting the error below while starting scrapy, although the spider works OK. Then I asked some teammates and they also get this error, while I don't.

```
2016-01-19 17:05:33 [boto] ERROR: Caught exception reading instance data
Traceback (most recent call last):
   File "/usr/lib/python2.7/site-packages/boto/utils.py", line 210, in
retry_url
     r = opener.open(req, timeout=timeout)
   File "/usr/lib64/python2.7/urllib2.py", line 431, in open
     response = self._open(req, data)
   File "/usr/lib64/python2.7/urllib2.py", line 449, in _open
     '_open', req)
   File "/usr/lib64/python2.7/urllib2.py", line 409, in _call_chain
     result = func(*args)
   File "/usr/lib64/python2.7/urllib2.py", line 1229, in http_open
     return self.do_open(httplib.HTTPConnection, req)
   File "/usr/lib64/python2.7/urllib2.py", line 1199, in do_open
     raise URLError(err)
URLError: <urlopen error timed out>
```

**Comment by nramirezuy** [comment]
@kmike Wasn't that solved here? https://github.com/scrapy/scrapy/pull/1421

@stummjr If you aren't using the `S3Downloader`, you can manually disable it.
Just add this to `settings.py`:
```python
DOWNLOAD_HANDLERS = {'s3': None}
```

**Comment by kmike** [comment]
@nramirezuy yes it was, but AFAIK it is not backported to Scrapy 1.0.x, so the fix will be a part of the next release (1.1).
